File: translation/4.5-haiku/test_3.go
package main

import (
	"fmt"
)

// Stringer definiert die String-Methode
type Stringer interface {
	String() string
}

// Validator definiert die IsValid-Methode
type Validator interface {
	IsValid() bool
}

// Entry erweitert sowohl Stringer als auch Validator
type Entry interface {
	Stringer
	Validator
}

// PrintValid gibt alle Items aus, die das Entry-Interface erfüllen
func PrintValid[T Entry](items []T) {
	for _, item := range items {
		if item.IsValid() {
			fmt.Println("✓", item.String())
		} else {
			fmt.Println("✗", item.String())
		}
	}
}

// --- Konkreter Typ ---

// Email implementiert das Entry-Interface
type Email struct {
	Address string
}

// NewEmail ist der Konstruktor für Email
func NewEmail(address string) *Email {
	return &Email{
		Address: address,
	}
}

// String implementiert das Stringer-Interface
func (e *Email) String() string {
	return e.Address
}

// IsValid implementiert das Validator-Interface
func (e *Email) IsValid() bool {
	return len(e.Address) > 3
}

func main() {
	emails := []*Email{
		NewEmail("alice@example.com"),
		NewEmail("ab"),
		NewEmail("bob@test.org"),
	}

	PrintValid(emails)
}

File: translation/4.5-haiku/test_3.ts
nterface Stringer {
	String(): string;
}

interface Validator {
	IsValid(): boolean;
}

interface Entry extends Stringer, Validator {}

function PrintValid<T extends Entry>(items: T[]): void {
	for (const item of items) {
		if (item.IsValid()) {
			console.log("✓", item.String());
		} else {
			console.log("✗", item.String());
		}
	}
}

// --- Konkreter Typ ---

class Email implements Entry {
	Address: string;

	constructor(address: string) {
		this.Address = address;
	}

	String(): string {
		return this.Address;
	}

	IsValid(): boolean {
		return this.Address.length > 3;
	}
}

const emails: Email[] = [
	new Email("alice@example.com"),
	new Email("ab"),
	new Email("bob@test.org"),
];

PrintValid(emails);